Source/WebCore/loader/FrameLoader.cpp

```diff
@@ -1272,6 +1272,28 @@ void FrameLoader::loadInSameDocument(URL url, RefPtr<SerializedScriptValue> stat
     RefPtr document = m_frame->document();
     // Update the data source's request with the new URL to fake the URL change
     URL oldURL = document->url();
+    bool hashChange = equalIgnoringFragmentIdentifier(url, oldURL) && !equalRespectingNullity(url.fragmentIdentifier(), oldURL.fragmentIdentifier());
+
+    if (document->settings().navigationAPIEnabled() && !url.isAboutBlank()) {
+        if (RefPtr domWindow = document->domWindow()) {
+            auto navigationType = determineNavigationType(policyChecker().loadType(), isNewNavigation);
+
+            // FIXME: This is likely not the ideal location for the navigate event and should happen earlier.
+            if (navigationType != NavigationNavigationType::Traverse) {
+                if (!domWindow->protectedNavigation()->dispatchPushReplaceReloadNavigateEvent(url, navigationType, oldURL))
+                    return;
```

Comment:
I do wonder if we should do some final steps before a return, like `started();` and `m_client->didFinishLoad();` below.

Reply:
That is a good point, but I think if the event is prevented it acts as a load abort, and I don't think the client has to be notified, since the load/fragment navigation has not actually started.